plugins/mongodb/mongodbembed.cpp

```diff
@@ -0,0 +1,2322 @@
+     * @param view Document View for looking up bound parameters from the function definition.
+     * @param start The point in the embedded script to start parsing.
+     */
+    template<typename T>
+    void MongoDBEmbedFunctionContext::buildArray(T &ctx, const bsoncxx::document::view &view, const char*& start)
+    {
+        std::string findStr = " ,(){}[]\t\n\"";
+        const char *end;
+
+        while (true) 
+        {
+            if (findStr.find(*start) == std::string::npos) 
+            {
+                end = start + 1;
+                while (*end != ',' && *end != ' ' && *end != ']') end++;
+                auto param = std::string(start, end - start);
+                bool isRsvd = !checkDoc(param, view);
+                if (!isRsvd) 
+                {
+                    param = param.substr(1);
+                    insertValueArr(ctx, view[param]);
+                }
+                else ctx << param;
+                start = end;
+            }
+            if (*start == '{') // Open subdocument 
+            {
+                ctx << open_document << [&](key_context<> kctx) {
+                    buildDocument(kctx, view, ++start);
+                } << close_document;
+            } 
+
+            if (*start == '[') // Open subarray 
+            {
+                ctx << open_array << [&](array_context<> actx) {
+                    buildArray(actx, view, ++start);
+                } << close_array;
+            }
+            if (*start == ']') 
+            {
+                start++;
+                break;
+            } 
+            else start++;
+        }
+    }
+
+    /**
+     * @brief Helper function for streaming Key Value pairs into a builder object. It is called on document and
+     * key_context objects. Array_context objects don't need Key Value pairs and can take any number of elements.
+     * 
+     * @tparam T Either a bsoncxx::stream::builder::document or bsoncxx::stream::builder::key_context
+     * @param builder Builder object for streaming elements into.
+     * @param view Document View for looking up bound parameters from the function definition.
+     * @param start The point in the embedded script to start parsing.
+     */
+    template<typename T>
+    void MongoDBEmbedFunctionContext::bindKVP(T &builder, const bsoncxx::document::view &view, const char *&start)
+    {
+        const char *end = start + 1;
+        while (*end != ':' && *end != '\"' && *end != ' ') end++; // pointer to end of key
+        auto key = std::string(start, end - start);
+        bool isRsvd = !checkDoc(key, view); // If this is a parameter of the function we need to set a flag
+        if (!isRsvd) key = key.substr(1); // Remove "$" for the key if it's in view
+
+        start = end + 1;
+        while (*start == ' ' || *start == ':') start++;
+
+        if (*start == '\"') 
+        {
+            end = ++start;
+            while (*end != '\"') end++;
```

Comment:
Is the input guaranteed to be wel-formed, or could this loop past the end of a invalid string?

Reply:
This is input from the embedded script, so there is a good chance that it will not be well-formed. For loops like this I can add something like while (*end && *end != '\"').